File: src/cpp/permutationaction.cpp
#include "permutationaction.h"

#include "common.h"

PermutationAction::PermutationAction()
	:m_permutation()
{
}

void PermutationAction::setPermutation(const Permutation &permutation)
{
	m_permutation = permutation;
}

PuzzleAction *PermutationAction::cloneAction()
{
	PermutationAction *clonedAction = new PermutationAction();
	clonedAction->m_permutation = m_permutation;
	return clonedAction;
}

void PermutationAction::execute(PuzzleData *puzzleData, bool inverted)
{
	// TODO: PermutationAction::execute()
	//if (m_permutation == null)
	//	reportError("No Permutation applied to BasicAction");
	//else
		m_permutation.apply(puzzleData, inverted);
}

void PermutationAction::reportError(const string &message)
{
	cout << "PermutationAction error: " << message << endl;
}

File: src/java/PermutationAction.java
package permgame;


public class PermutationAction implements PuzzleAction
{
	private Permutation m_permutation;
	private boolean     m_inverted;
	
	public PermutationAction()
	{
		m_permutation = null;
		m_inverted = false;
	}
	
	public void setPermutation(Permutation permutation)
	{
		m_permutation = permutation;
	}

	@Override
	public void setInverted(boolean inverted)
	{
		m_inverted = inverted;
	}
	
	/*
	 * (non-Javadoc)
	 *
	 * @see permgame.PuzzleAction#cloneAction()
	 */
	@Override
	public PuzzleAction cloneAction()
	{
		PermutationAction cloneAction = new PermutationAction();
		cloneAction.m_permutation = m_permutation.clone();
		cloneAction.m_inverted = m_inverted;
		return cloneAction;
	}
	
	/*
	 * (non-Javadoc)
	 *
	 * @see permgame.PuzzleAction#execute(java.util.ArrayList,
	 * java.util.ArrayList)
	 */
	@Override
	public void execute(PuzzleData puzzleData, boolean inverted)
	{
		if (m_permutation == null)
			reportError("No permutation applied to BasicAction");
		else
			m_permutation.apply(puzzleData, inverted);
	}

	private void reportError(String message)
	{
		System.err.println("Error: " + message);
	}
}
